[Vimuth Mendis] iP

vimuthm opened the pull request:
# GFMDPro

> “Unpopular Opinion: `Hello Peter` saved the spiderman trailer” – Vimuth Mendis ([source](https://github.com/nus-cs2103-AY2122S1/ip/pull/61))

## Unnecessary plug

Check out this [user guide](https://ideas-lab-nus.github.io/vizpro/) to verify my gfmd skills 😎 
Give the [tool](https://ideas-lab-nus.github.io/vizpro/playground/) a go while you're at it :) (Excuse the missing icons, the google bucket in use was taken down, sigh)

## Moving on...
Duke removes the need to remember things you need to do as it will occupy all your time! It's, 

 - anxiety-inducing
 - **fun**
 - time-consuming

All you need to not do is, 

1. work
2. any and all work you have
3. till the week ends and the grader counts 0 for participation ☠️

And it's your ~~procrastination~~ *laid-back work ethic* to blame!

My Todos:

 - [ ] 2103 iP, tP
 - [ ] 2106 lab, tut
 - [ ] ST, MA 20 hours of lecs
 - [x]  These checkboxes are unnecessary, it'll never end :")

---

Let me show you how to reverse a linked list for the lols:

```java
Node reverse(Node node) {
    Node prev = null;
    Node current = node;
    Node next = null;
    while (current != null) {
        next = current.next;
        current.next = prev;
        prev = current;
        current = next;
     }
     node = prev;
     return node;
}
```
 courtesy of the cs student community's trusted [GeeksForGeeks](https://www.geeksforgeeks.org/reverse-a-linked-list/)

Commit:
Add Gradle support

vimuthm committed:
Add greet, echo, exit

vimuthm committed:
Add listing functionality

vimuthm committed:
Add task completion

vimuthm committed:
Implement toString method for Task

vimuthm committed:
Add sub tasks

vimuthm committed:
Improve splitting

vimuthm committed:
Add input.txt and expected.txt for testing

vimuthm committed:
Add error handling

vimuthm committed:
Add testing for errors

vimuthm committed:
Add delete functionality

vimuthm committed:
Add command enum

vimuthm committed:
Add save and load features

vimuthm committed:
Add date time handling

vimuthm committed:
Merge branch 'branch-Level-7'

vimuthm committed:
Handle date time in saving/loading

vimuthm committed:
Add Ui, Storage, Parser, TaskList and Command Classes

vimuthm committed:
Create packages

vimuthm committed:
Add JUnit testing

vimuthm committed:
Remove wildcard import

vimuthm committed:
Create JAR File

vimuthm committed:
Add javadocs

vimuthm committed:
Tweak code style to meet standards

vimuthm committed:
Add find command

vimuthm committed:
Merge branch 'branch-A-JavaDoc'

vimuthm committed:
Resolve merge conflict

vimuthm committed:
Resolve merge conflict

Cyn7hius commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,126 @@
+package duke;
+
+import duke.command.Command;
+import duke.command.Action;
+import duke.command.List;
+import duke.command.Add;
+import duke.command.End;
+import duke.command.Find;
+
+import duke.exception.DukeException;
+import duke.exception.MissingDescriptionException;
+import duke.exception.UnknownCommandException;
+import duke.exception.InvalidFormatException;
+import duke.exception.InvalidDateTimeException;
+
+import duke.tasks.Deadline;
+import duke.tasks.Event;
+import duke.tasks.Task;
+import duke.tasks.ToDo;
+
+import java.time.LocalDate;
+import java.time.LocalTime;
```
I think it might be easier to read if the import statements are in alphabetical order?

vimuthm replied:
True. I'll send in a fix soon, thanks for pointing this out!

Cyn7hius commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,128 @@
+     *
+     * @return size as an int
+     */
+    public int size() {
```
I think this would be clearer if it was a verb like `getSize()` instead of just `size()`

vimuthm replied:
Yes, this violates the standards. Will fix!

Cyn7hius commented on src/main/java/duke/command/Find.java:
```diff
@@ -0,0 +1,32 @@
+import duke.exception.InvalidFormatException;
+import duke.tasks.Task;
+
+public class Find extends Command {
```
I think there should be a javaDocs for this class

vimuthm replied:
Will add in!

Cyn7hius reviewed (commented):
Wow super good OOP and very clean code!😄

vimuthm committed:
Fix issues raised in PR review

vimuthm committed:
Merge branch 'add-gradle-support' into branch-A-Gradle

NoraYUuu commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,132 @@
+    public ArrayList<Task> load() throws DukeException {
+        String home = System.getProperty("user.dir");
+        Path path = Paths.get(home, this.filepath);
+        boolean directoryExists = Files.exists(path);
```
Maybe can name as isDirectoryFound for it to look more like a boolean?

vimuthm replied:
> Overall, namings are easy to read and understand, good job!
> You may want to add more javadoc for some public methods for reader to understand them better.

Thank you for the useful input! I've made the necessary changes.

NoraYUuu commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,132 @@
+                        }
+                        break;
+                    case "D":
+                        LocalDate date2 = LocalDate.parse(data[4]);
```
Maybe can name as deadlineDate and eventDate for easy reference, as date and date2 can be confusing?

NoraYUuu commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,82 @@
+            + "\t\t `delete ${i}` to delete task i\n"
+            + "\t\t `find ${keyword}` to find tasks by keyword\n"
+            + "\t\t `bye` to end this chat\n";
+    private static final String endMessage = "\n\tSad to see you go :(\n\t...shutting down...";
```
Final and static value may be named as END_MESSAGE.

NoraYUuu commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Shows error messages.
+     */
+    public void showError(String error) {
```
I like how you clearly differentiate different messages and named them in this class!

NoraYUuu commented on src/main/java/duke/command/Action.java:
```diff
@@ -0,0 +1,70 @@
+        this.words = words;
+    }
+
+    @Override
```
You may want to add a javadoc for this public method since it is quite important~

NoraYUuu commented on src/main/java/duke/exception/SavingError.java:
```diff
@@ -0,0 +1,7 @@
+package duke.exception;
+
+public class SavingError extends DukeException {
```
I like how you separate exception classes into subgroups to make the structure even cleaner

NoraYUuu reviewed (commented):
Overall, namings are easy to read and understand, good job!
You may want to add more javadoc for some public methods for reader to understand them better.

vimuthm committed:
Fix issues raised in 2nd PR review

vimuthm committed:
Add build and test with gradle

vimuthm committed:
Merge branch 'branch-A-Gradle'

vimuthm committed:
Add checkstyle

CSYGalvin commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,124 @@
+        return c;
+    }
+
+    private static Task handleFormat(String[] split, String message, Type type) throws DukeException {
```
Maybe you can include javadocs for this method since it is non-trivial?

CSYGalvin commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,128 @@
+    /**
+     * Saves the provided TaskList in an appropriate format.
+     *
+     * @param taskList TaskList which is to be saved.
```
Maybe you can include a `@throws` line for the javadocs?

CSYGalvin commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+    /**
+     * Returns Task at given index.
+     *
+     * @param x int index
```
Looks clean! Just missed a punctuation mark here.

CSYGalvin commented on src/main/java/duke/command/Command.java:
```diff
@@ -0,0 +1,31 @@
+    public abstract void execute(TaskList taskList, Ui ui, Storage storage) throws DukeException;
+
+    /**
+     * Returns if the command is an exit.
```
Do you mean returns true if the command is an exit?

CSYGalvin commented on src/main/java/duke/command/End.java:
```diff
@@ -0,0 +1,21 @@
+/**
+ * Class that handles the Bye command
+ */
+public class End extends Command {
```
It might be confusing if the class name is `End` but it handles the `Bye` command.

CSYGalvin reviewed (commented):
Good Job!! I really enjoyed reading your project and learning from it as well

mslevis commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,132 @@
+
+    /**
+     * Returns a String for Find command.
+     *
```
Perhaps you can include a params tag for keyword?

mslevis commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,82 @@
+    /**
+     * Shows end message.
+     */
+    public void sayBye() {
```
Perhaps you can name the method showBye to be consistent with the other ui methods?

vimuthm committed:
Add a GUI to the application

vimuthm committed:
Address checkstyle issues; Add javadocs

vimuthm committed:
Merge branch 'branch-Level-10'

vimuthm committed:
Add assertions

Introducing assertions to the codebase

Let's add assertions to,
* Make sure the interceptor has a latest response
* Ensure the target is non-null in the invocation handler
* Check that the type parameter to the Action class is 0/1

vimuthm committed:
Refactor to increase code quality

Increase code quality according to standards set in the module

Let's,
* Add descriptive comments where needed
* Introduce guard clauses
* Remove arrow-head like code structures
* Introduce meaningful variable names
* Remove duplicate code

vimuthm committed:
Add stream usage for string formation

Introduce streams for string concatenation from task list

Let's use streams to,
* Build up the string instead of StringBuilder in stringify,
  along with a Functional Interface
* Build a string for the find command, using filter and map

vimuthm committed:
Merge pull request #1 from VimuthM/branch-A-Assertions

Add assertions

vimuthm committed:
Merge branch 'master' into branch-A-CodeQuality

vimuthm committed:
Resolve conflict

vimuthm committed:
Merge pull request #2 from VimuthM/branch-A-CodeQuality

Refactor to increase code quality

vimuthm committed:
Merge branch 'master' into branch-A-Streams

vimuthm committed:
Merge pull request #3 from VimuthM/branch-A-Streams

Add stream usage for string formation

vimuthm committed:
Create gradle.yml

vimuthm committed:
Fix checkstyle errors

vimuthm committed:
Merge branch 'master' of https://github.com/VimuthM/ip

vimuthm committed:
Fix checkstyle errors

vimuthm committed:
Fix checkstyle issues in test files

vimuthm committed:
Add sort functionality

vimuthm committed:
Merge branch 'branch-BCD-Extension'

vimuthm committed:
Enhance the GUI

vimuthm committed:
Complete javadocs

vimuthm committed:
Add User Guide
